x/delayedack/rollapp_hooks.go

```diff
@@ -32,79 +36,147 @@ func (im IBCMiddleware) FinalizeRollappPackets(ctx sdk.Context, rollappID string
+		"num packets", len(rollappPendingPackets))
+
 	for _, rollappPacket := range rollappPendingPackets {
-		logger.Debug("Finalizing IBC rollapp packet", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel(), "type", rollappPacket.Type)
-		var err error
-		switch rollappPacket.Type {
-		case commontypes.RollappPacket_ON_RECV:
-			logger.Debug("Calling OnRecvPacket", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel())
-			wrappedFunc := func(ctx sdk.Context) error {
-				ack := im.IBCModule.OnRecvPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
-				// If async, return
-				if ack == nil {
-					return nil
-				}
-				// If sync, check if the acknowledgement is successful
-				if !ack.Success() {
-					return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, string(ack.Acknowledgement()))
-				}
-				// Write the acknowledgement to the chain only if it is synchronous
-				_, chanCap, err := im.keeper.LookupModuleByChannel(ctx, rollappPacket.Packet.DestinationPort, rollappPacket.Packet.DestinationChannel)
-				if err != nil {
-					return err
-				}
-				err = im.keeper.WriteAcknowledgement(ctx, chanCap, rollappPacket.Packet, ack)
-				if err != nil {
-					return err
-				}
-				return nil
-			}
-			err = osmoutils.ApplyFuncIfNoError(ctx, wrappedFunc)
-			if err != nil {
-				logger.Error("writing acknowledgement", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel(), "error", err.Error())
-			}
-		case commontypes.RollappPacket_ON_ACK:
-			logger.Debug("Calling OnAcknowledgementPacket", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel())
-			wrappedFunc := func(ctx sdk.Context) error {
-				err := im.IBCModule.OnAcknowledgementPacket(ctx, *rollappPacket.Packet, rollappPacket.Acknowledgement, rollappPacket.Relayer)
-				if err != nil {
-					return err
-				}
-				return nil
-			}
-			err := osmoutils.ApplyFuncIfNoError(ctx, wrappedFunc)
-			if err != nil {
-				logger.Error("calling OnAcknowledgementPacket", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel(), "error", err.Error())
-			}
-		case commontypes.RollappPacket_ON_TIMEOUT:
-			logger.Debug("Calling OnTimeoutPacket", "rollappID", rollappID, "sequence", rollappPacket.Packet.GetSequence(), "destination channel", rollappPacket.Packet.GetDestChannel())
-			wrappedFunc := func(ctx sdk.Context) error {
-				err := im.IBCModule.OnTimeoutPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
-				if err != nil {
-					return err
-				}
-				return nil
-			}
-			err := osmoutils.ApplyFuncIfNoError(ctx, wrappedFunc)
+		if rollappPacket.RollappId != rollappID {
```

Comment:
I think much more efficeint method would be to only iterate on the relevant rollapp packets from the first place vs iterating everything and filter after.
This can be done by expanding the `ListRollappPacketByStatus` to get an optional rollapp_id arguement or in another way. The rollapp key is built in such way that it should allow it as it's built using `/status/rollapp-id`

Reply:
@omritoptix 
It's actually built like this, so `status/proof_height/rollapp_id/...`

```
result := append(statusPrefix, keySeparatorBytes...)
result = append(result, proofHeightBytes...)
result = append(result, keySeparatorBytes...)
result = append(result, rollappIdBytes...)
result = append(result, keySeparatorBytes...)
result = append(result, packetUIDBytes...)
```
so we get a key like this: `“\0\x02/\0\0\0\0\0\0k\b/rolintdev_5832-2/channel-1\0\0\0\0\0\0\0\x14”`

I think we would need to also put the proof height in the prefix, or just rearrange the parts of the key.